typing/typecore.ml

```diff
@@ -381,6 +386,17 @@ let extract_label_names env ty =
 let is_principal ty =
   not !Clflags.principal || get_level ty = generic_level
 
+(* Returns [Some ty_elt] if the expected type can be used for type-directed
+   disambiguation of an array literal. *)
+let disambiguate_array_literal env expected_ty =
+  if is_principal expected_ty then
+    if is_floatarray_type env expected_ty then
```

Comment:
I assume this is written like that (and not `if is_principal expected_ty && is_floatarray_type env expected_ty then Some ... else None`) to support, later, other kinds of arrays. Is that right?

In particular, it could already be useful to support `bytes` (and possibly `string`) literals. Would that require a lot of extra changes in the backend?